### PR TITLE
[zh-cn]: Update translation of Object.prototype.constructor

### PR DESCRIPTION
**Pull request by JasonLamv-t**
### Description

Such the title.

### Motivation

Ditto

### Additional details

None.

### Related issues and pull requests

None.

### CODE REVIEW

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 1]
```diff
@@ -5,29 +5,51 @@ slug: Web/JavaScript/Reference/Global_Objects/Object/constructor
...
+const a2 = new Array();
+a2.constructor === Array; // true
 
-const a = new Array
-a.constructor === Array // true
+const n = 3;
+n.constructor === Number; // true
+```
+
+Note that `constructor` usually comes from the constructor's [`prototype`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) property. If you have a longer prototype chain, you can usually expect every object in the chain to have a `constructor` property.
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:43]:
```suggestion
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 2]
```diff
@@ -5,29 +5,51 @@ slug: Web/JavaScript/Reference/Global_Objects/Object/constructor
...
+
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:44]:
```suggestion
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 3]
```diff
@@ -5,29 +5,49 @@ slug: Web/JavaScript/Reference/Global_Objects/Object/constructor
...
+对创建该实例对象的构造函数的引用。
+
+{{js_property_attributes(1, 0, 1)}}
+
+> **备注：** 这个属性默认会在每个构造函数的 [`prototype`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) 属性上创建，并由该构造函数创建的所有对象继承。
 
 ## 描述
 
 所有对象（使用 `Object.create(null)` 创建的对象除外）都将具有 `constructor` 属性。在没有显式使用构造函数的情况下，创建的对象（例如对象和数组文本）将具有 `constructor` 属性，这个属性指向该对象的基本对象构造函数类型。
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:23]:
```suggestion
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 4]
```diff
@@ -5,29 +5,49 @@ slug: Web/JavaScript/Reference/Global_Objects/Object/constructor
...
 
+除了 [`null` 原型对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object#null_原型对象)之外，任何对象都会在其 `[[Prototype]]` 上有一个 `constructor` 属性。使用字面量创建的对象也会有一个指向该对象构造函数类型的 `constructor` 属性，例如，数组字面量创建的 {{jsxref("Array")}} 对象和[对象字面量](/zh-CN/docs/Web/JavaScript/Reference/Operators/Object_initializer)创建普通对象。
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:24]:
```suggestion
除了 [`null` 原型对象](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object#null_原型对象)之外，任何对象都会在其 `[[Prototype]]` 上有一个 `constructor` 属性。使用字面量创建的对象也会有一个指向该对象构造函数类型的 `constructor` 属性，例如，数组字面量创建的 {{jsxref("Array")}} 对象和[对象字面量](/zh-CN/docs/Web/JavaScript/Reference/Operators/Object_initializer)创建的普通对象。
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 5]
```diff
@@ -38,166 +58,243 @@ n.constructor === Number // true
...
-  this.name = name
+  this.name = name;
 }
 ```
 
-### 改变对象的 constructor
+### 为对象的 constructor 属性赋值
 
-可以为除了 `null` 和 `undefined`（因为这两者没有相应的构造函数）之外的任何类型指定 `constructor` 属性（如 `String`、`Number`、`Boolean` 等），但基本类型不会保留这些更改（也不会抛出异常）。也是同样的原因，基本类型允许设置任何属性（除了 `null` 和 `undefined`），而不会产生副作用。就是说，每当把这样的基本类型当成对象使用时，其对应的构造函数的实例就会在语句执行后立即被创建和丢弃。
+可以为非原型对象的 constructor 属性赋值。
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:78]:
```suggestion
可以为非基本类型封装对象的 constructor 属性赋值。
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 6]
```diff
@@ -38,166 +58,243 @@ n.constructor === Number // true
...
+arr instanceof String; // false
+arr instanceof Array; // true
 ```
 
-### 改变函数的 constructor
+`constructor` 属性没有受到保护，可以被重新赋值或被覆盖，因此在检测变量类型时，通常应避免使用它，而应该使用更不易出错的方法，如对于对象使用 `instanceof` 和 [`Symbol.toStringTag`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag)，对于基本类型使用 [`typeof`](/zh-CN/docs/Web/JavaScript/Reference/Operators/typeof)。
+
+### 更改构造函数原型对象的 constructor 属性
+
+每个构造函数都有一个 [`prototype`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) 属性，当通过 [`new`](/zh-CN/docs/Web/JavaScript/Reference/Operators/new) 运算符调用时，该属性将成为实例的 `[[Prototype]]`。因此，`ConstructorFunction.prototype.constructor` 将成为实例的 `[[Prototype]]` 上的属性，如上述所示。
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:118]:
```suggestion
每个构造函数都有一个 [`prototype`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) 属性，当通过 [`new`](/zh-CN/docs/Web/JavaScript/Reference/Operators/new) 运算符调用时，该属性将成为实例的 `[[Prototype]]`。因此，`ConstructorFunction.prototype.constructor` 将成为实例的 `[[Prototype]]` 上的属性，如前面所述。
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 7]
```diff
@@ -38,166 +58,243 @@ n.constructor === Number // true
...
-CreatedConstructor.prototype = Object.create(Parent.prototype)
+CreatedConstructor.prototype = Object.create(Parent.prototype);
 
-CreatedConstructor.prototype.create = function create() {
-  return new this.constructor()
+CreatedConstructor.prototype.create = function () {
+  return new this.constructor();
+};
+
+new CreatedConstructor().create().create(); // TypeError: new CreatedConstructor().create().create is undefined, since constructor === Parent
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:155]:
```suggestion
new CreatedConstructor().create().create(); // TypeError: new CreatedConstructor().create().create is undefined，因为 constructor === Parent
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 8]
```diff
@@ -38,166 +58,243 @@ n.constructor === Number // true
...
+```
+
+在上面的示例中，会抛出一个异常，因为 `constructor` 链接到 `Parent`。为了避免这种情况，只需分配你将要使用的必要构造函数即可。
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:158]:
```suggestion
在上面的示例中，会抛出一个异常，因为 `constructor` 链接到 `Parent`。为了避免这种情况，只需分配你将要使用的必需构造函数即可。
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 9]
```diff
@@ -38,166 +58,243 @@ n.constructor === Number // true
...
+    value: Child,
+    enumerable: false,
+    writable: true,
+    configurable: true,
+  },
+});
+
+Child.prototype.getOffsetByInitialPosition = function () {
+  const position = this.position;
+  // 使用 `this.constructor`，希望 `getStartPosition` 存在于一个静态方法中。
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:234]:
```suggestion
  // 使用 `this.constructor`，以期 `getStartPosition` 存在于一个静态方法中。
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 10]
```diff
@@ -38,166 +58,243 @@ n.constructor === Number // true
...
-    offsetY: startPosition.y - position.y
-  }
-}
+    offsetY: startPosition.y - position.y,
+  };
+};
+
+new Child(1, 1).getOffsetByInitialPosition();
+// Error: this.constructor.getStartPosition is undefined, since the
+// constructor is Child, which doesn't have the getStartPosition static method
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:245]:
```suggestion
// Error: this.constructor.getStartPosition is undefined，
// 因为构造函数是 Child，它没有 getStartPosition 静态方法
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 11]
```diff
@@ -38,166 +56,243 @@ n.constructor === Number // true
...
+CreatedConstructor.prototype.create = function () {
+  return new this.constructor();
+};
 
-CreatedConstructor.prototype.create = function create() {
-  return new this.constructor()
+new CreatedConstructor().create().create(); // TypeError: new CreatedConstructor().create().create is undefined，因为 constructor === Parent
+```
+
+在上面的示例中，会抛出一个异常，因为 `constructor` 链接到 `Parent`。为了避免这种情况，只需分配你将要使用的必需构造函数即可。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:156]:
```suggestion
在上面的示例中，会抛出一个异常，因为 `constructor` 链接到 `Parent`。为了避免这种情况，只需赋值为你将要使用的必要的构造函数即可。
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 12]
```diff
@@ -38,166 +56,243 @@ n.constructor === Number // true
...
+    offsetY: startPosition.y - position.y,
+  };
+};
+
+console.log(new Child(1, 1).getOffsetByInitialPosition()); // { offsetX: -1, offsetY: -1 }
 ```
 
+使用 `Object.setPrototypeOf()` 可能会对性能产生不利影响，因此请确保它仅在必要时使用，并在构造函数声明后立即使用，并在创建任何实例之前使用，以避免对象被“污染”。
+
 > **备注：** 设置或更新构造函数可能会导致结果不同且令人困惑的结果。为了防止它，只需在特定情况下定义 `constructor`。多数情况，不使用 `constructor`，并且不需要重新定义。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:296]:
```suggestion
> **备注：** 设置或更新构造函数可能会导致结果不同且令人困惑的结果。为了防止它，只需在特定情况下定义 `constructor`。多数情况，不使用 `constructor`，并且不需要重新对其赋值。
```

**Review by yin1999 — APPROVED** [review]
几个建议，也可以不修改

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md [hunk 13]
```diff
@@ -5,29 +5,47 @@ slug: Web/JavaScript/Reference/Global_Objects/Object/constructor
...
+const a2 = new Array();
+a2.constructor === Array; // true
 
-const n = new Number(3)
-n.constructor === Number // true
+const n = 3;
+n.constructor === Number; // true
+```
+
+请注意，`constructor` 属性通常来自构造函数的 [`prototype`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) 属性。如果你有一个更长的原型链，通常可以期望链中的每个对象都有一个 `constructor` 属性。
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:41]:
```suggestion
请注意，`constructor` 属性通常来自构造函数的 [`prototype`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype) 属性。如果你有一个更长的原型链，通常可以假定链中的每个对象都有一个 `constructor` 属性。
```

File: files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md (hunk 11 above)
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/object/constructor/index.md:156]:
```suggestion
在上面的示例中，会抛出一个异常，因为 `constructor` 链接到 `Parent`。为了避免这种情况，只需将其赋值为你将要使用的必要构造函数即可。
```

**Comment by yin1999** [comment]
Thank you @JasonLamv-t!

**Pull request merged**